contentcuration/contentcuration/utils/recommendations.py

```diff
@@ -61,67 +76,307 @@ def create_backend(self) -> Backend:
+            request_hash = self._generate_request_hash(request)
+            override_threshold = self._extract_override_threshold(request)
+            data = list(RecommendationsCache.objects
+                        .filter(request_hash=request_hash, override_threshold=override_threshold)
+                        .order_by('override_threshold', 'rank')
+                        .values('contentnode_id', 'rank'))
+            if len(data) > 0:
+                return EmbeddingsResponse(data=data)
+            else:
+                return None
         except Exception as e:
-            return EmbeddingsResponse(error=e)
+            logging.exception(e)
+            return None
+
+    def _generate_request_hash(self, request) -> str:
+        """
+        Generates a unique hash for a given request. It serializes the request
+        attributes that make it unique, then generates a hash of this serialization.
+
+        To prevent cache duplication, the hash is generated
+        independent of the override_threshold parameter.
+
+        :param request: The request for which to generate a unique hash.
+        :return: A unique hash representing the request
+        """
 
-    def embed_content(self, nodes: List[ContentNode]) -> EmbeddingsResponse:
+        params_copy = request.params.copy() if request.params else {}
+        params_copy.pop('override_threshold', None)
 
+        unique_attributes = json.dumps({
+            'params': params_copy,
+            'json': request.json,
+        }, sort_keys=True).encode('utf-8')
+
+        return hashlib.md5(unique_attributes).hexdigest()
+
+    def cache_embeddings_request(self, request: BackendRequest, response: BackendResponse) -> bool:
+        """
+        Caches the recommendations request and response. It performs a bulk insert of the
+        recommendations into the RecommendationsCache table, ignoring any conflicts.
+
+        :param request: The request to cache.
+        :param response: The response to cache.
+        :return: A boolean indicating whether the caching was successful.
+        :rtype: bool
+        """
+
+        try:
+            nodes = self._extract_data(response)
+            request_hash = self._generate_request_hash(request)
+            existing_cache = set(RecommendationsCache.objects.filter(request_hash=request_hash)
+                                 .values_list('contentnode_id', flat=True))
+            override_threshold = self._extract_override_threshold(request)
+            new_cache = [
+                RecommendationsCache(
+                    request_hash=request_hash,
+                    contentnode_id=node['contentnode_id'],
+                    rank=node['rank'],
+                    override_threshold=override_threshold,
+                ) for node in nodes if node['contentnode_id'] not in existing_cache
+            ]
+            RecommendationsCache.objects.bulk_create(new_cache, ignore_conflicts=True)
```

Comment:
Theoretically, having `ignore_conflicts` should mean you can omit checking `existing_cache`, because:
> On databases that support it (all but Oracle), setting the ignore_conflicts parameter to True tells the database to ignore failure to insert any rows that fail constraints such as duplicate unique values. ([ref](https://docs.djangoproject.com/en/5.0/ref/models/querysets/#bulk-create))

So with the `unique_together` expression on `request_hash` and `contentnode_id`, then it should ignore failing to create those.

Reply:
It would be great to have a unit test to verify the previously described behavior.

Reply:
I wrote tests for the cache that confirmed the above when `ignore_conflicts` is `True` or `False`. The only difference is that when `ignore_conflicts=False`, an exception is raised, brining the creation to a halt. So I think we can safely remove the  `existing_cache` check as long as `ignore_conflicts=True`.

Reply:
Nice! sounds perfect